File: server/product-service/src/main/kotlin/com/sudoo/productservice/controller/dto/UpsertProductDto.kt
package com.sudoo.productservice.controller.dto

import java.time.LocalDateTime

data class UpsertProductDto(
    val productId: String? = null,
    val sku: String? = null,
    val name: String? = null,
    val description: String? = null,
    val price: Float? = null,
    val listedPrice: Float? = null,
    val amount: Int? = null,
    val soldAmount: Int? = null,
    val discount: Int? = null,
    val startDateDiscount: LocalDateTime? = null,
    val endDateDiscount: LocalDateTime? = null,
    val sellable: Boolean? = null,
    val images: List<String>? = null,
    val supplierId: String? = null,
    val categoryIds: List<String>? = null,
)

File: client/web/sudoo/lib/data/dto/discovery/upsert_product_dto.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:sudoo/domain/model/discovery/upsert_product.dart';

part 'upsert_product_dto.g.dart';

@JsonSerializable()
class UpsertProductDto {
  final String productId;
  final String sku;
  final String name;
  final String description;
  final double price;
  final double listedPrice;
  final int amount;
  final int soldAmount;
  final int discount;
  final DateTime? startDateDiscount;
  final DateTime? endDateDiscount;
  final bool saleable;
  final List<String> images;
  final List<String> categoryIds;

  const UpsertProductDto(
    this.productId,
    this.sku,
    this.name,
    this.description,
    this.price,
    this.listedPrice,
    this.amount,
    this.soldAmount,
    this.discount,
    this.startDateDiscount,
    this.endDateDiscount,
    this.saleable,
    this.images,
    this.categoryIds,
  );

  factory UpsertProductDto.fromJson(Map<String, dynamic> json) =>
      _$UpsertProductDtoFromJson(json);

  Map<String, dynamic> toJson() => _$UpsertProductDtoToJson(this);
}

extension Mapper on UpsertProductDto {
  UpsertProduct toUpsertProduct() {
    return UpsertProduct(
      productId: productId,
      sku: sku,
      name: name,
      description: description,
      price: price,
      listedPrice: listedPrice,
      amount: amount,
      soldAmount: soldAmount,
      discount: discount,
      startDateDiscount: startDateDiscount,
      endDateDiscount: endDateDiscount,
      saleable: saleable,
      images: images,
      categoryIds: categoryIds,
    );
  }
}
